File: Stack/evaluate-reverse-polish-notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        res_li = []

        def apply(ele, res_li):
            op2 = int(res_li.pop())
            op1 = int(res_li.pop())

            # print('in apply:',op1, op2)

            if ele == "+":
                res_li.append(op1+ op2)
            elif ele == "-":
                res_li.append(op1-op2)
            elif ele == "*":
                res_li.append(op1*op2)
            else:
                res_li.append(int(op1/op2))
            # print(res_li)

        def add_ele(ele, res_li):
            res_li.append(ele)
            # print(res_li)

        # reverse the tokens 

        for ele in tokens:
            # print(' for ele:', ele)
            if ele in ("+", "-", "*",  "/"):
                apply(ele, res_li)
            else:
                add_ele(ele, res_li)
        
        # print(res_li)
        return int(res_li.pop())

File: Stack/evaluate-reverse-polish-notation.java
class Solution {
    public int evalRPN(String[] tokens) {
        int stack[] = new int[tokens.length];
        int s_i = 0;

        int op1, op2, res;

        for (String ele:  tokens){
            if (ele.equals("+") || ele.equals("-") || ele.equals("*") || ele.equals("/")){
                op1 = stack[s_i-2];
                op2 = stack[s_i -1];
                s_i-=2;

                if (ele.equals("+"))
                    res = op1+op2;
                else if(ele.equals("-"))
                    res = op1- op2;
                else if (ele.equals("*"))
                    res = op1 *op2;
                else
                    res = (int)op1 / op2;
                // System.out.println(" the res is: "+ res);
                stack[s_i]= res;
                s_i++;
            }
            else{
                // System.out.println(" the else case token is: "+ ele);
                stack[s_i]= Integer.parseInt(ele);
                s_i++;
            }
        }
        return stack[0];
    }
}